gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/resources/auth/handler/impl/ClientAuthHandlerImpl.java

```diff
@@ -73,6 +86,25 @@ public void handle(RoutingContext routingContext) {
 
                 // the client might has been upgraded after authentication process, get the new value
                 Client authenticatedClient = authHandler.result();
+
+                // get SSL certificate thumbprint to bind with access token
+                try {
+                    SSLSession sslSession = routingContext.request().sslSession();
+                    if (sslSession != null) {
+                        Certificate[] peerCertificates = sslSession.getPeerCertificates();
+                        X509Certificate peerCertificate = (X509Certificate) peerCertificates[0];
+                        routingContext.put(ConstantKeys.PEER_CERTIFICATE_THUMBPRINT, getThumbprint(peerCertificate, "SHA-256"));
+                    } else if (sslSession == null && (authenticatedClient.isTlsClientCertificateBoundAccessTokens() || domain.usePlainFapiProfile())) {
+                        routingContext.fail(new InvalidClientException("Missing or invalid peer certificate"));
+                        return;
+                    }
+                } catch (SSLPeerUnverifiedException | CertificateEncodingException | NoSuchAlgorithmException ce ) {
+                    if (authenticatedClient.isTlsClientCertificateBoundAccessTokens() || domain.usePlainFapiProfile()) {
+                        routingContext.fail(new InvalidClientException("Missing or invalid peer certificate"));
+                        return;
+                    }
+                }
+
```

Comment:
I wonder if it's better to put the ConstantKeys.PEER_CERTIFICATE_THUMBPRINT in the Client object instead of re-do this check.